kmk/modules/encoder.py

```diff
@@ -0,0 +1,221 @@
+                self.encoder_direction = 'R'
+            elif new_encoder_state == (self.eps.ON, self.eps.ON):  # Turned left 1
+                if self.encoder_direction == 'L':
+                    self.encoder_value = self.encoder_value - 1
+        elif self.encoder_state == (self.eps.OFF, self.eps.ON):  # R3 or L1
+            if new_encoder_state == (self.eps.OFF, self.eps.OFF):  # Turned left 1
+                self.encoder_direction = 'L'
+            elif new_encoder_state == (self.eps.ON, self.eps.ON):  # Turned right 1
+                if self.encoder_direction == 'R':
+                    self.encoder_value = self.encoder_value + 1
+        else:  # self.encoder_state == '11'
+            if new_encoder_state == (self.eps.ON, self.eps.OFF):  # Turned left 1
+                self.encoder_direction = 'L'
+            elif new_encoder_state == (self.eps.OFF, self.eps.ON):  # Turned right 1
+                self.encoder_direction = 'R'
+            elif new_encoder_state == (
+                self.eps.ON,
+                self.eps.ON,
+            ):  # Skipped intermediate 01 or 10 state, however turn completed
+                if self.encoder_direction == 'L':
+                    self.encoder_value = self.encoder_value - 1
+                elif self.encoder_direction == 'R':
+                    self.encoder_value = self.encoder_value + 1
+
+        self.encoder_state = new_encoder_state
+
+        if self.vel_mode:
+            self.vel_ts = ticks_ms()
+
+        if self.encoder_state != self.last_encoder_state:
+            self.position_change = self.invert_rotation(
+                self.encoder_value, self.last_encoder_value
+            )
+
+            self.last_encoder_state = self.encoder_state
+            self.last_encoder_value = self.encoder_value
+
+            if self.position_change > 0:
+                # return self.increment_key
+                return 0
+            elif self.position_change < 0:
+                # return self.decrement_key
+                return 1
+            else:
+                return None
+
+    # invert knob direction if encoder pins are soldered backwards
+    def invert_rotation(self, new, old):
+        if self.is_inverted:
+            return -(new - old)
+        else:
+            return new - old
+
+    # returns knob velocity as milliseconds between position changes(detents)
+    def vel_report(self):
+        self.encoder_speed = self.vel_ts - self.last_vel_ts
+        self.last_vel_ts = self.vel_ts
+        return self.encoder_speed
+
+
+class EncoderHandler(Module):
+
+    encoders = []
+    debug_enabled = False # not working as inttended, do not use for now
+
+    def __init__(self, pad_a, pad_b, encoder_map):
+        self.pad_a = pad_a
+        self.pad_b = pad_b
+        self.encoder_count = len(self.pad_a)
+        self.encoder_map = encoder_map
+        self.make_encoders()
+
+    def on_runtime_enable(self, keyboard):
+        raise NotImplementedError
```

Comment:
These should be blank `return` as opposed to raising an error if they are intentionally not used. This is to catch you if you flat out forget to include them.